File: web/src/lib/agent-smith/tick.js
// =====================================================================
//  Agent Smith — the tick-loop.
//  What a compiled agent does each tick: observe → diagnose → commit.
//  The agent acts on the STATE OF THE SOLUTION, not on internal reasoning.
//
//  observe : read a slice of the shared solution-state Omega (tier 0 cheap
//            probe; escalate to a model read only on residue). Never stores
//            the task; re-reads Omega. → residual gap + a candidate act.
//  diagnose: read the residual descent → a typed limit
//            (converging | compute-limited | structure-limited/stall).
//  commit  : a RUNTIME-owned gate reading the OUTCOME of the candidate act.
//            Fires iff (sufficiency: outcome-gain clears the price) AND
//            (coherence: preserves the agent's coherence). Else the agent
//            stays a pure observer — valid, almost free, contributes 0.
//
//  The model (providers.think) is used only at the irreducible seam: the
//  domain judgment inside observe/commit. Everything else is deterministic.
// =====================================================================

import { waterFill } from "./identity";
import { think } from "./providers";

export const LIMIT = {
  CONVERGING: "converging",
  COMPUTE: "compute-limited",
  STRUCTURE: "structure-limited",
};

export const OUTCOME = {
  COMMIT: "commit",
  OBSERVE: "observe", // gate did not fire; pure observer this tick
  DECLINE: "decline", // off-purpose / incoherent for this interaction
  QUIESCENT: "quiescent", // task-agent reached its attainable outcome
};

/**
 * observe(agent, omega, ctx)
 * Tier-0 read is deterministic: measure the residual gap of THIS agent's
 * purpose against the current solution-state via ctx.probe. If the slice is
 * not enough to determine an act (gap ambiguous), escalate to a model read
 * (Tier 1). Returns { residual, candidate, tier, slice }.
 */
async function observe(agent, omega, ctx) {
  // Tier 0: cheap deterministic probe of the solution-state.
  const slice = ctx.probe(omega, agent); // { residual, hint }
  let residual = slice.residual;
  let tier = 0;
  let candidate = null;

  // The EFFECTIVE gain of a scene scales with the residual gap: acting pays
  // in proportion to how far the solution-state is from this agent's purpose.
  // At the floor the gap is ~0, so no scene clears the price → the agent
  // observes (and a task-agent quiesces). This is the single mechanism behind
  // both the pure observer and task halting: the gate reads the state, and
  // when the state does not call for the agent, it does not act.
  const price = waterFillPrice(agent);
  const gap = Math.max(0, residual - agent.floorNorm);
  const effGain = (s) => s.gain.g0 * gap;
  const active = agent.scenes.filter((s) => effGain(s) > price);
  if (active.length === 0) {
    // no scene clears the price for this state → nothing to do (observer)
    return { residual, candidate: null, tier, slice, price };
  }

  // Ambiguity test: escalate to a model read (Tier 1) only when the gap is in
  // the "unsure" band near the floor, where a judgment is genuinely needed;
  // otherwise form the candidate deterministically from the highest-gain scene.
  const unsureBand = gap > 0 && gap < 0.15;
  if (unsureBand && ctx.useModel) {
    tier = 1;
    const j = await judgeSufficiency(agent, slice, ctx);
    residual = j.residual ?? residual;
    candidate = j.act ? { ...j.act, expectedGain: effGain(agent.scenes[0]) } : null;
  } else {
    // deterministic candidate: the highest-gain active scene
    const scene = active.reduce((a, b) => (effGain(b) > effGain(a) ? b : a));
    candidate = { scene: scene.name, hook: scene.hook, expectedGain: effGain(scene) };
  }
  return { residual, candidate, tier, slice, price };
}

/**
 * diagnose(agent) — read the residual descent over recent ticks.
 * Uses agent.stallWindow (recent deltas). Returns a LIMIT.
 */
function diagnose(agent) {
  const win = agent.stallWindow;
  if (win.length === 0) return LIMIT.COMPUTE;
  const recent = win.slice(-3);
  const sum = recent.reduce((a, b) => a + b, 0);
  if (sum <= 1e-9) return LIMIT.STRUCTURE; // no descent over the window: stall
  const last = win[win.length - 1];
  if (last > 0.02) return LIMIT.CONVERGING;
  return LIMIT.COMPUTE;
}

/**
 * commit(agent, candidate, residual, omega, ctx) — the runtime sufficiency
 * gate. Reads the OUTCOME of the candidate act. Fires iff sufficiency AND
 * coherence. On fire: deposit act, increment count, extend trajectory, and
 * change omega. Returns { outcome, delta, act }.
 */
async function commit(agent, obs, omega, ctx) {
  const { candidate, residual, price } = obs;

  // task-agent that has reached its attainable outcome → quiescent, halts —
  // but only if it actually did work to get there (committed ≥ 1 act). A
  // task-agent that starts already at its floor never had work: it is a pure
  // OBSERVER (ready to act if the shared state raises its residual), not done.
  if (agent.regime === "task" && residual <= agent.floorNorm + 1e-6) {
    if (agent.count > 0) {
      agent.state = "quiescent";
      return { outcome: OUTCOME.QUIESCENT, delta: 0, act: null };
    }
    agent.state = "observing";
    return { outcome: OUTCOME.OBSERVE, delta: 0, act: null };
  }

  // no candidate cleared the price → pure observer (floor case). Free.
  if (!candidate) {
    agent.state = "observing";
    return { outcome: OUTCOME.OBSERVE, delta: 0, act: null };
  }

  // sufficiency: does the candidate's OUTCOME-gain clear the attention price?
  // (release at sufficiency for the parent, not at the agent's own floor.)
  const gain = candidate.expectedGain;
  const sufficient = gain > price;

  // coherence: would the act break the agent's coherence condition?
  const coherent = ctx.coherent ? ctx.coherent(agent, candidate) : true;

  if (!sufficient || !coherent) {
    agent.state = coherent ? "observing" : "running";
    return {
      outcome: coherent ? OUTCOME.OBSERVE : OUTCOME.DECLINE,
      delta: 0,
      act: null,
      reason: !sufficient ? "below price" : "incoherent",
    };
  }

  // FIRE. Perform the domain work via the hook (model at the irreducible
  // seam) and read its outcome. The hook is opaque; we read only its effect.
  let hookResult = null;
  if (ctx.useModel && ctx.runHook) {
    hookResult = await ctx.runHook(agent, candidate, obs, omega);
  }
  // outcome effect on the residual (deterministic descent scaled by gain,
  // or the model-reported reduction if the hook returned one)
  const reduction =
    hookResult && typeof hookResult.reduction === "number"
      ? hookResult.reduction
      : Math.min(residual - agent.floorNorm, 0.15 + 0.1 * (gain / (price + 1e-9) - 1));
  const before = agent.disposition;
  agent.disposition = Math.max(agent.floorNorm, before - Math.max(0, reduction));

  // I2: increment the monotone committed count; extend the trajectory.
  agent.count += 1;
  agent.trajectory.push(labelStep(candidate, agent.count));
  agent.lastResidual = agent.disposition;
  agent.state = "running";

  const delta = before - agent.disposition;
  // update omega: the committed act changes the shared solution-state that
  // every other agent will read next tick (outcome-space coordination).
  ctx.applyToOmega?.(omega, agent, candidate, delta, hookResult);

  return {
    outcome: OUTCOME.COMMIT,
    delta,
    act: {
      scene: candidate.scene,
      hook: candidate.hook,
      count: agent.count,
      content: hookResult?.content ?? null,
      model: hookResult?.model ?? null,
    },
  };
}

/**
 * tick(agent, omega, ctx) — one full observe → diagnose → commit cycle,
 * respecting phase exclusion (observe/diagnose in construction; commit in
 * commitment). Returns a record for the UI.
 */
export async function tick(agent, omega, ctx) {
  if (agent.state === "quiescent") {
    return { agent: agent.name, outcome: OUTCOME.QUIESCENT, count: agent.count };
  }

  // construction phase: observe + diagnose (no act)
  agent.phase = "construction";
  const obs = await observe(agent, omega, ctx);
  const limit = diagnose(agent);

  // commitment phase: the gate
  agent.phase = "commitment";
  const res = await commit(agent, obs, omega, ctx);

  // record the residual delta for the next diagnosis
  agent.stallWindow.push(res.delta || 0);
  if (agent.stallWindow.length > 8) agent.stallWindow.shift();

  return {
    agent: agent.name,
    regime: agent.regime,
    outcome: res.outcome,
    limit,
    tier: obs.tier,
    price: Number(obs.price?.toFixed?.(3) ?? obs.price ?? 0),
    residual: Number(agent.disposition.toFixed(3)),
    delta: Number((res.delta || 0).toFixed(3)),
    count: agent.count,
    scene: res.act?.scene ?? obs.candidate?.scene ?? null,
    content: res.act?.content ?? null,
    model: res.act?.model ?? null,
    reason: res.reason ?? null,
  };
}

// ---- helpers --------------------------------------------------------

/** Normalised floor: the residual level the agent treats as "at purpose". */
export function initFloorNorm(agent) {
  // scale the graph floor into the [0,1] residual space; a small positive
  // number so the gap never closes to zero (the floor theorem).
  agent.floorNorm = Math.min(0.08, 0.02 + 0.01 * (agent.floor || 1));
  agent.disposition = 1.0;
  agent.lastResidual = 1.0;
  return agent;
}

function waterFillPrice(agent) {
  const scenes = agent.scenes.map((s) => ({ id: s.id, g0: s.gain.g0, invMarginal: s.gain.invMarginal }));
  const { price } = waterFill(scenes, agent.budget);
  return price;
}

function labelStep(candidate, count) {
  // the trajectory ADDRESS: how we got here (scene + count), not the content.
  return `${candidate.scene}#${count}`;
}

// Tier-1 model judgment of sufficiency: ask a model whether the current
// slice is enough to act, and how much residual an act would remove.
async function judgeSufficiency(agent, slice, ctx) {
  const prompt =
    `You are the sufficiency gate for an agent whose purpose is "${agent.purpose.target}".\n` +
    `Current solution-state slice: ${JSON.stringify(slice.hint ?? {})}.\n` +
    `Residual gap to purpose: ${slice.residual.toFixed(3)} (floor ${agent.floorNorm.toFixed(3)}).\n` +
    `Answer with ONE word: ACT if committing an act now would advance the purpose, ` +
    `or WAIT if the state does not yet call for this agent.`;
  try {
    const r = await think(prompt, { keys: ctx.keys, crowd: ctx.crowd, crowdSize: ctx.crowdSize });
    const act = /\bACT\b/i.test(r.text);
    return {
      act: act ? { scene: agent.scenes[0].name, hook: agent.scenes[0].hook, expectedGain: agent.scenes[0].gain.g0, model: r.model } : null,
      residual: slice.residual,
    };
  } catch {
    // model unavailable → fall back to deterministic candidate
    return { act: { scene: agent.scenes[0].name, hook: agent.scenes[0].hook, expectedGain: agent.scenes[0].gain.g0 }, residual: slice.residual };
  }
}

File: crates/agent-smith/src/tick.rs
// =====================================================================
//  Agent Smith — the tick-loop.
//  What a compiled agent does each tick: observe → diagnose → commit.
//  The agent acts on the STATE OF THE SOLUTION, not on internal reasoning.
//  Port of tick.js (the JS async is gone: the only reason it was async was
//  the model call, which is now the synchronous Hook seam).
//
//  observe : read a slice of the shared solution-state Omega (tier 0 cheap
//            probe; escalate to a model read only on residue). Never stores
//            the task; re-reads Omega. → residual gap + a candidate act.
//  diagnose: read the residual descent → a typed limit.
//  commit  : a RUNTIME-owned gate reading the OUTCOME of the candidate act.
//            Fires iff (sufficiency: outcome-gain clears the price) AND
//            (coherence). Else the agent stays a pure observer — valid,
//            almost free, contributes 0.
// =====================================================================

use crate::compile::{Agent, AgentState, Phase};
use crate::hook::{Candidate, HookResult, Slice};
use crate::identity::{water_fill_price, SceneCost};
use crate::town::{Ctx, Omega};

/// The diagnosed limit on the agent's descent this tick.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum Limit {
    Converging,
    #[serde(rename = "compute-limited")]
    Compute,
    #[serde(rename = "structure-limited")]
    Structure,
}

/// The outcome of a tick's commitment gate.
#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Outcome {
    Commit,
    Observe,   // gate did not fire; pure observer this tick
    Decline,   // off-purpose / incoherent for this interaction
    Quiescent, // task-agent reached its attainable outcome
}

impl Outcome {
    pub fn as_str(&self) -> &'static str {
        match self {
            Outcome::Commit => "commit",
            Outcome::Observe => "observe",
            Outcome::Decline => "decline",
            Outcome::Quiescent => "quiescent",
        }
    }
}

/// The per-tick record surfaced to the UI/CLI.
#[derive(Debug, Clone, serde::Serialize)]
pub struct TickRecord {
    pub agent: String,
    pub regime: Option<String>,
    pub outcome: Outcome,
    pub limit: Option<Limit>,
    pub tier: u8,
    pub price: f64,
    pub residual: f64,
    pub delta: f64,
    pub count: u64,
    pub scene: Option<String>,
    pub content: Option<String>,
    pub model: Option<String>,
    pub reason: Option<String>,
}

/// The result of the observe phase.
struct Obs {
    residual: f64,
    candidate: Option<Candidate>,
    tier: u8,
    slice: Slice,
    price: f64,
}

/// The result of the commit gate.
struct CommitResult {
    outcome: Outcome,
    delta: f64,
    act: Option<Act>,
    reason: Option<String>,
}

struct Act {
    scene: String,
    #[allow(dead_code)]
    hook: String,
    #[allow(dead_code)]
    count: u64,
    content: Option<String>,
    model: Option<String>,
}

/// Normalised floor: the residual level the agent treats as "at purpose".
/// Scale the graph floor into [0,1] residual space; a small positive number so
/// the gap never closes to zero (the floor theorem).
pub fn init_floor_norm(agent: &mut Agent) {
    agent.floor_norm = (0.02 + 0.01 * if agent.floor > 0.0 { agent.floor } else { 1.0 }).min(0.08);
    agent.disposition = 1.0;
    agent.last_residual = 1.0;
}

fn water_fill_price_for(agent: &Agent) -> f64 {
    let scenes: Vec<SceneCost> = agent
        .scenes
        .iter()
        .map(|s| SceneCost { id: s.id.clone(), gain: s.gain })
        .collect();
    water_fill_price(&scenes, agent.budget)
}

/// observe: tier-0 deterministic probe; escalate to a model read only in the
/// "unsure" band near the floor, where a judgment is genuinely needed.
fn observe(agent: &mut Agent, omega: &Omega, ctx: &Ctx) -> Obs {
    let slice = ctx.probe(omega, agent);
    let mut residual = slice.residual;
    let mut tier = 0u8;
    let mut candidate: Option<Candidate> = None;

    let price = water_fill_price_for(agent);
    let gap = (residual - agent.floor_norm).max(0.0);
    // effective gain of a scene scales with the residual gap: acting pays in
    // proportion to how far the solution-state is from this agent's purpose.
    let eff_gain = |g0: f64| g0 * gap;

    let active: Vec<&crate::compile::CompiledScene> =
        agent.scenes.iter().filter(|s| eff_gain(s.gain.g0()) > price).collect();
    if active.is_empty() {
        // no scene clears the price for this state → nothing to do (observer)
        return Obs { residual, candidate: None, tier, slice, price };
    }

    // ambiguity: escalate to a model read only in the unsure band near floor.
    let unsure_band = gap > 0.0 && gap < 0.15;
    if unsure_band && ctx.use_model {
        tier = 1;
        let act = ctx.hook.judge_sufficiency(agent, &slice);
        if act {
            let s0 = &agent.scenes[0];
            candidate = Some(Candidate {
                scene: s0.name.clone(),
                hook: s0.hook.clone(),
                expected_gain: eff_gain(agent.scenes[0].gain.g0()),
            });
        }
        // residual unchanged in the deterministic-judgment fallback
        residual = slice.residual;
    } else {
        // deterministic candidate: the highest-gain active scene
        let scene = active
            .iter()
            .copied()
            .reduce(|a, b| if eff_gain(b.gain.g0()) > eff_gain(a.gain.g0()) { b } else { a })
            .unwrap();
        candidate = Some(Candidate {
            scene: scene.name.clone(),
            hook: scene.hook.clone(),
            expected_gain: eff_gain(scene.gain.g0()),
        });
    }

    Obs { residual, candidate, tier, slice, price }
}

/// diagnose: read the residual descent over recent ticks (agent.stall_window).
fn diagnose(agent: &Agent) -> Limit {
    let win = &agent.stall_window;
    if win.is_empty() {
        return Limit::Compute;
    }
    let recent: f64 = win.iter().rev().take(3).sum();
    if recent <= 1e-9 {
        return Limit::Structure; // no descent over the window: stall
    }
    let last = *win.last().unwrap();
    if last > 0.02 {
        Limit::Converging
    } else {
        Limit::Compute
    }
}

/// commit: the runtime sufficiency gate. Reads the OUTCOME of the candidate
/// act. Fires iff sufficiency AND coherence. On fire: deposit act, increment
/// count (I2), extend trajectory, and change omega.
fn commit(agent: &mut Agent, obs: &Obs, omega: &mut Omega, ctx: &Ctx) -> CommitResult {
    let residual = obs.residual;
    let price = obs.price;

    // task-agent that has reached its attainable outcome → quiescent, halts —
    // but only if it actually did work to get there (committed >= 1 act). A
    // task-agent that starts already at its floor never had work: it is a pure
    // OBSERVER (ready to act if the shared state raises its residual).
    if agent.regime == crate::typecheck::Regime::Task && residual <= agent.floor_norm + 1e-6 {
        if agent.count > 0 {
            agent.state = AgentState::Quiescent;
            return CommitResult { outcome: Outcome::Quiescent, delta: 0.0, act: None, reason: None };
        }
        agent.state = AgentState::Observing;
        return CommitResult { outcome: Outcome::Observe, delta: 0.0, act: None, reason: None };
    }

    // no candidate cleared the price → pure observer (floor case). Free.
    let candidate = match &obs.candidate {
        None => {
            agent.state = AgentState::Observing;
            return CommitResult { outcome: Outcome::Observe, delta: 0.0, act: None, reason: None };
        }
        Some(c) => c.clone(),
    };

    // sufficiency: does the candidate's OUTCOME-gain clear the attention price?
    let gain = candidate.expected_gain;
    let sufficient = gain > price;

    // coherence: would the act break the agent's coherence condition?
    let coherent = ctx.coherent(agent, &candidate);

    if !sufficient || !coherent {
        agent.state = if coherent { AgentState::Observing } else { AgentState::Running };
        return CommitResult {
            outcome: if coherent { Outcome::Observe } else { Outcome::Decline },
            delta: 0.0,
            act: None,
            reason: Some(if !sufficient { "below price" } else { "incoherent" }.to_string()),
        };
    }

    // FIRE. Perform the domain work via the hook (model at the irreducible
    // seam) and read its outcome. The hook is opaque; we read only its effect.
    let hook_result: Option<HookResult> = ctx.run_hook(agent, &candidate, &obs.slice);

    // outcome effect on the residual (deterministic descent scaled by gain, or
    // the model-reported reduction if the hook returned one).
    let reduction = match hook_result.as_ref().and_then(|h| h.reduction) {
        Some(r) => r,
        None => {
            let det = 0.15 + 0.1 * (gain / (price + 1e-9) - 1.0);
            (residual - agent.floor_norm).min(det)
        }
    };
    let before = agent.disposition;
    agent.disposition = (before - reduction.max(0.0)).max(agent.floor_norm);

    // I2: increment the monotone committed count; extend the trajectory.
    agent.count += 1;
    agent.trajectory.push(label_step(&candidate, agent.count));
    agent.last_residual = agent.disposition;
    agent.state = AgentState::Running;

    let delta = before - agent.disposition;
    // update omega: the committed act changes the shared solution-state every
    // other agent reads next tick (outcome-space coordination).
    ctx.apply_to_omega(omega, agent, &candidate, delta, &hook_result);

    let content = hook_result.as_ref().and_then(|h| h.content.clone());
    let model = hook_result.as_ref().and_then(|h| h.model.clone());
    CommitResult {
        outcome: Outcome::Commit,
        delta,
        act: Some(Act {
            scene: candidate.scene.clone(),
            hook: candidate.hook.clone(),
            count: agent.count,
            content,
            model,
        }),
        reason: None,
    }
}

/// tick: one full observe → diagnose → commit cycle, respecting phase
/// exclusion (observe/diagnose in construction; commit in commitment).
pub fn tick(agent: &mut Agent, omega: &mut Omega, ctx: &Ctx) -> TickRecord {
    if agent.state == AgentState::Quiescent {
        return TickRecord {
            agent: agent.name.clone(),
            regime: Some(agent.regime.as_str().to_string()),
            outcome: Outcome::Quiescent,
            limit: None,
            tier: 0,
            price: 0.0,
            residual: round3(agent.disposition),
            delta: 0.0,
            count: agent.count,
            scene: None,
            content: None,
            model: None,
            reason: None,
        };
    }

    // construction phase: observe + diagnose (no act)
    agent.phase = Phase::Construction;
    let obs = observe(agent, omega, ctx);
    let limit = diagnose(agent);

    // commitment phase: the gate
    agent.phase = Phase::Commitment;
    let res = commit(agent, &obs, omega, ctx);

    // record the residual delta for the next diagnosis
    agent.stall_window.push(res.delta);
    if agent.stall_window.len() > 8 {
        agent.stall_window.remove(0);
    }

    let scene = res
        .act
        .as_ref()
        .map(|a| a.scene.clone())
        .or_else(|| obs.candidate.as_ref().map(|c| c.scene.clone()));

    TickRecord {
        agent: agent.name.clone(),
        regime: Some(agent.regime.as_str().to_string()),
        outcome: res.outcome,
        limit: Some(limit),
        tier: obs.tier,
        price: round3(obs.price),
        residual: round3(agent.disposition),
        delta: round3(res.delta),
        count: agent.count,
        scene,
        content: res.act.as_ref().and_then(|a| a.content.clone()),
        model: res.act.as_ref().and_then(|a| a.model.clone()),
        reason: res.reason,
    }
}

fn label_step(candidate: &Candidate, count: u64) -> String {
    // the trajectory ADDRESS: how we got here (scene + count), not the content.
    format!("{}#{}", candidate.scene, count)
}

fn round3(v: f64) -> f64 {
    (v * 1000.0).round() / 1000.0
}
